man/yokadid.1
.TH YOKADID 1 "July 10, 2009"
.SH NAME
yokadid \- commandline todo system
.SH SYNOPSIS
.B yokadid
.RI [ options ]...
.br
.SH DESCRIPTION
.B yokadid
is a Yokadi daemon that remind you due tasks. If you want to be automatically
reminded of due tasks, you can use the Yokadi daemon. The Yokadi daemon can be
launched via desktop autostart services.

In most desktop environments, you just need to create a symbolic link to yokadid
(or a shell script that calls it) in $HOME/.config/autostart/.

ln \-s \`which yokadid\` $HOME/.config/autostart/

.PP
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-\-datadir=<DIR>
Database directory.
.TP
.B \-k, \-\-kill
Kill Yokadi Daemon (you can specify database with \-db if you run multiple
Yokadid).
.TP
.B \-f, \-\-foreground
Don't fork background. Useful for debug.
.TP
.B \-h, \-\-help
Show summary of options and exit.
.SH SEE ALSO
.BR yokadi (1).
.br
.SH SEE ALSO
Website: http://yokadi.github.io

Mailing List: http://sequanux.org/cgi-bin/mailman/listinfo/ml-yokadi
.SH AUTHOR
yokadi was written by Aurélien Gâteau <mail@agateau.com> and Sébastien Renard <sebastien.renard@digitalfox.org>.
.PP
This manual page was written by Kartik Mistry <kartik@debian.org>,
for the Debian project (and may be used by others).


NEWS
v1.2.0 unreleased
- New features:
    - The new `p_merge` command lets you merge a project into another.
    - It is now possible to turn a task into a note with `t_to_note` and a note into a task with `n_to_task`.
- Bug fixes:
    - The `k_remove` command no longer ignores unused keywords.
    - HTML output has been fixed to no longer output strings wrapped in `b""`.
    - `t_list` filtering has been fixed so that `t_list --urgency 0` filters out tasks with a negative urgency, as expected.
- Improvements:
    - HTML output has been refreshed:
        - It looks more modern now.
        - Some fields have been removed (doneDate, creationDate).
        - The title, keywords and description fields have been merged.
        - An ID field has been added (handy to run a command on a task listed in the output).
        - Columns now use human-friendly titles.
- Misc:
    - The `--db` option is now deprecated and replaced by the `--datadir` option. `--db` will be removed in the next version.
    - Similarly, the `YOKADI_DB` environment variable is now deprecated and will be removed in the next version.
    - Yokadi no longer supports cryptography: encrypted databases will be decrypted at update.

v1.1.1 2016/11/11

- Improvements:
    - When listing multiple projects, order them alphabetically.
- Bug fixes:
    - Fixed parse error if the user sets a time of "17m".
    - When the user edits a tasks with t_edit and removes a keyword, remove the keyword from the task.
    - Made recurrence code work with dateutil 2.6.0.

v1.1.0 2016/09/03

- New features & Improvements:
    - A new command has been added: `t_medit`. `t_medit` lets you edit all tasks of a project in one go.
    - Aliases can now be modified. The name of the alias can be modified with `a_edit_name` and the command with `a_edit_command`.
    - Database format updates are now easier to run: just run `yokadi -u`, no more separate `update.py` command. Updates are also much faster.
    - Task lists have been improved:
        - Borders look nicer.
        - Some bugs in the rendering of the title column have been fixed (wrong width, badly cropped text).
    - Yokadi now uses standard paths by default: the database is stored in ~/.local/share/yokadi/yokadi.db and non-essential data is in ~/.cache/yokadi/.
    - Reviewed and improved documentation. Moved developer documentation to a separate dir (doc/dev).
- Bug fixes:
    - The code handling recurrences has been made more robust.
    - Recurrences are now stored in a more future proof way.
    - Fixed `bug_edit` crash.
    - Fixed negative keyword filter: A task with two keywords k1 and k2 would not be excluded by a filter !k1.

v1.0.2 2016/03/28

- Use a more portable way to get the terminal size. This makes it possible to use Yokadi inside Android terminal emulators like Termux
- Sometimes the task lock used to prevent editing the same task description from multiple Yokadi instances were not correctly released
- Deleting a keyword from the database caused a crash when a t_list returned tasks which previously contained this keyword

v1.0.1 2015/12/03

- User changes:
    - Make sure installing via pip installs the required dependencies

- Developer changes:
    - Improved release process

v1.0.0 2015/11/29

- User changes:
    - Fixed an issue which caused t_list to fail when filtering by keywords on large lists
    - Removed the project keywords feature. It was not very useful and made the searching code more complicated
    - Fixed ical support: it now works with ical 3.6 or later
    - Improved documentation
    - Added Keywords field to yokadi.desktop
- Developer changes:
    - Yokadi has been ported to Python 3
    - The application now uses SQLAlchemy instead of SQLObject to access the SQLite database

v0.14 2014/05/03
- Command changes:
    - t_add, n_add:
        - Allow creating two tasks with the same title (useful for recurrent tasks, like "buy bread").
        - Allow using _ to select last project, making it possible to do multiple t_add on the same project with `t_add _ <task description>`.
        - Add --describe option to start describing the task right after adding it.
    - t_describe, n_describe:
        - Safer task description editing: task is updated each time the editor saves, a lock manager now prevents multiple edits.
        - Use .md suffix instead of .txt for the temporary filename to allow some smart things with editors that understand Markdown.
        - Use project and task name for the temporary filename. Useful when using graphical editors or when your terminal title shows the current running command.
    - t_due:
        - When called with a time argument which is before current time, set due date to the day after.
    - t_show:
        - Show the task ID.
    - t_list:
        - Use month and year for the task age if the task is older than 12 months.
        - Add support for arbitrary minimum date for --done.
        - Fixed broken help.
    - n_list:
        - Display creation date instead of age.
        - Notes are now grouped by date.
    - p_list:
        - Show task count per project.
    - p_remove:
        - Show the number of associated tasks in the prompt.
    - p_edit:
        - Handle case where user tries to rename a project using the name of an existing project.

- yokadid:
    - Add --restart option and --log option.
    - Set process name with setproctitle.
    - Configuration keys can now be overridden using environment variables.

- Misc:
    - Date/time commands now support %d/%m/%y date format.
    - Replaced xyokadi with a desktop file.
    - Updated README to match real output.

- Developer specific changes:
    - Command parser has been ported from optparse to argparse.
    - Code is now PEP 8 compliant, with the exception of camelCase usage.
    - All imports have been changed to absolute imports (ie `import yokadi.<something>`).
    - Code has been reorganized into different sub directories.

yokadi/yokadid.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
"""
Yokadi daemon. Used to monitor due tasks and warn user.

@author: Sébastien Renard <sebastien.renard@digitalfox.org>
@license: GPL v3 or later
"""

import sys
import os
import time
from datetime import datetime, timedelta
from signal import SIGTERM, SIGHUP, signal
from subprocess import Popen
from argparse import ArgumentParser

from yokadi.core import fileutils

try:
    import setproctitle
except ImportError:
    print("You don't have the setproctitle package.")
    print("Get it on http://pypi.python.org/pypi/setproctitle/")
    print("Or use 'easy_install setproctitle'")
    sys.exit(1)

from yokadi.core.daemon import Daemon
from yokadi.core import basepaths
from yokadi.yical.yical import YokadiIcalServer

from yokadi.core import db
from yokadi.core.db import Project, Task, getConfigKey
from yokadi.ycli import commonargs


# Daemon polling delay (in seconds)
PROCESS_INTERVAL = 30
EVENTLOOP_INTERVAL = 1

# Ical daemon default port
DEFAULT_TCP_ICAL_PORT = 8000

# Event sender to main loop
event = [True, ""]


def sigTermHandler(signal, stack):
    """Handler when yokadid receive SIGTERM"""
    print("Receive SIGTERM. Exiting")
    print("End of yokadi Daemon")
    event[0] = False
    event[1] = "SIGTERM"


def sigHupHandler(signal, stack):
    """Handler when yokadid receive SIGHUP"""
    print("Receive SIGHUP. Reloading configuration")
    event[0] = False
    event[1] = "SIGHUP"


def eventLoop():
    """Main event loop"""
    delta = timedelta(hours=float(getConfigKey("ALARM_DELAY")))
    suspend = timedelta(hours=float(getConfigKey("ALARM_SUSPEND")))
    cmdDelayTemplate = getConfigKey("ALARM_DELAY_CMD")
    cmdDueTemplate = getConfigKey("ALARM_DUE_CMD")
    session = db.getSession()
    # For the two following dict, task id is key, and value is (duedate, triggerdate)
    triggeredDelayTasks = {}
    triggeredDueTasks = {}
    activeTaskFilter = [Task.status != "done",
                        Task.projectId == Project.id,
                        Project.active == True]  # noqa

    def process(now):
        delayTasks = session.query(Task).filter(Task.dueDate < now + delta,
                                                Task.dueDate > now,
                                                *activeTaskFilter)
        dueTasks = session.query(Task).filter(Task.dueDate < now,
                                              *activeTaskFilter)
        processTasks(delayTasks, triggeredDelayTasks, cmdDelayTemplate, suspend)
        processTasks(dueTasks, triggeredDueTasks, cmdDueTemplate, suspend)

    nextProcessTime = datetime.today().replace(microsecond=0)
    while event[0]:
        now = datetime.today().replace(microsecond=0)
        if now > nextProcessTime:
            process(now)
            nextProcessTime = now + timedelta(seconds=PROCESS_INTERVAL)
        time.sleep(EVENTLOOP_INTERVAL)


def processTasks(tasks, triggeredTasks, cmdTemplate, suspend):
    """Process a list of tasks and trigger action if needed
    @param tasks: list of tasks
    @param triggeredTasks: dict of tasks that has been triggered. Dict can be updated
    @param cmdTemplate: command line template to execute if task trigger
    @param suspend: timedelta beetween to task trigger"""
    now = datetime.now()
    for task in tasks:
        if task.id in triggeredTasks and triggeredTasks[task.id][0] == task.dueDate:
            # This task with the same dueDate has already been triggered
            if now - triggeredTasks[task.id][1] < suspend:
                # Task has been trigger recently, skip to next
                continue
        print("Task %s is due soon" % task.title)
        cmd = cmdTemplate.replace("{ID}", str(task.id))
        cmd = cmd.replace("{TITLE}", task.title.replace('"', '\"'))
        cmd = cmd.replace("{PROJECT}", task.project.name.replace('"', '\"'))
        cmd = cmd.replace("{DATE}", str(task.dueDate))
        process = Popen(cmd, shell=True)
        process.wait()
        # TODO: redirect stdout/stderr properly to Log (not so easy...)
        triggeredTasks[task.id] = (task.dueDate, datetime.now())


def killYokadid(pidFile):
    """Kill Yokadi daemon
    @param pidFile: file where the pid of the daemon is stored
    """
    # reuse Daemon.stop() code
    daemon = Daemon(pidFile)
    daemon.stop()


def parseOptions(defaultPidFile, defaultLogFile):
    parser = ArgumentParser()

    commonargs.addArgs(parser)

    parser.add_argument("-i", "--icalserver",
                        dest="icalserver", default=False, action="store_true",
                        help="Start the optional HTTP Ical Server")

    parser.add_argument("-p", "--port",
                        dest="tcpPort", default=DEFAULT_TCP_ICAL_PORT,
                        help="TCP port of ical server (default: %s)" % DEFAULT_TCP_ICAL_PORT,
                        metavar="PORT")

    parser.add_argument("-l", "--listen",
                        dest="tcpListen", default=False, action="store_true",
                        help="Listen on all interface (not only localhost) for ical server")

    parser.add_argument("-k", "--kill",
                        dest="kill", default=False, action="store_true",
                        help="Kill the Yokadi daemon. The daemon is found from the process ID stored in the file"
                             " specified with --pid")

    parser.add_argument("--restart",
                        dest="restart", default=False, action="store_true",
                        help="Restart the Yokadi daemon. The daemon is found from the process ID stored in the file"
                             " specified with --pid")

    parser.add_argument("-f", "--foreground",
                        dest="foreground", default=False, action="store_true",
                        help="Don't fork background. Useful for debug")

    parser.add_argument("--pid",
                        dest="pidFile", default=defaultPidFile,
                        help="File in which Yokadi daemon stores its process ID (default: %s)" % defaultPidFile)

    parser.add_argument("--log",
                        dest="logFile", default=defaultLogFile,
                        help="File in which Yokadi daemon stores its log output (default: %s)" % defaultLogFile)

    return parser.parse_args()


class YokadiDaemon(Daemon):
    def __init__(self, dbPath, options):
        Daemon.__init__(self, options.pidFile, stdout=options.logFile, stderr=options.logFile)
        self.dbPath = dbPath
        self.options = options

    def run(self):
        db.connectDatabase(self.dbPath, createIfNeeded=False)
        print("Using %s" % self.dbPath)
        session = db.getSession()

        # Basic tests :
        if not len(session.query(db.Config).all()) >= 1:
            print("Your database seems broken or not initialised properly. Start yokadi command line tool to do it")
            sys.exit(1)

        # Start ical http handler
        if self.options.icalserver:
            yokadiIcalServer = YokadiIcalServer(self.options.tcpPort, self.options.tcpListen)
            yokadiIcalServer.start()

        # Start the main event Loop
        try:
            while event[1] != "SIGTERM":
                eventLoop()
                event[0] = True
        except KeyboardInterrupt:
            print("\nExiting...")


def main():
    # TODO: check that yokadid is not already running for this database ? Not very harmful...
    # Set process name to "yokadid"
    setproctitle.setproctitle("yokadid")

    # Make the event list global to allow communication with main event loop
    global event

    defaultPidFile = os.path.join(basepaths.getRuntimeDir(), "yokadid.pid")
    defaultLogFile = os.path.join(basepaths.getLogDir(), "yokadid.log")
    args = parseOptions(defaultPidFile, defaultLogFile)
    _, dbPath = commonargs.processArgs(args)

    if args.kill:
        killYokadid(args.pidFile)
        sys.exit(0)

    if args.pidFile == defaultPidFile:
        fileutils.createParentDirs(args.pidFile, mode=0o700)

    if args.logFile == defaultLogFile:
        fileutils.createParentDirs(args.logFile, mode=0o700)

    signal(SIGTERM, sigTermHandler)
    signal(SIGHUP, sigHupHandler)

    if args.restart:
        daemon = YokadiDaemon(dbPath, args)
        daemon.restart()

    daemon = YokadiDaemon(dbPath, args)
    if args.foreground:
        daemon.run()
    else:
        daemon.start()


if __name__ == "__main__":
    main()
